### ISSUE TITLE
HigherRangeOfVision.js help

### ISSUE THREAD

**Issue by Poshwosh** [cursey/kanan#74]
I'm trying to create a script which modifies the range of vision beyond its limit, similar to Step29's mod_sharker's mod (since I couldn't restore it...)

I started searching for unknown value on Cheat Engine, constantly changing between minimum and maximum vision of range. I came into the conclusion minimum is 5000, maximum is 15000.
There are a few of those values, and they refuse not to work alone, as in it just reverts back to the original, except the second one. It seems to modify and affect all of them at once, so this is what needs to be modified.

This change was very simple. We search for `C8 42 00 60 6A 46 00 00 34 42 00 00 D2` to find that second address I was talking about. However, that values constantly changes between the login screen, character selection screen, Uldah and Iria, so we need to add wildcards. `6E 00 00 00 00 00 00 C8 42 00 ?? ?? ?? 00 00 34 42 00 00` works just fine. We replace those wildcards with `50 43 47` to achieve 50000 vision of range.

Now the problem is that, when you change channels, the value resets to your default configuration (15000). Need to break the function that reapplies the value. I right clicked the address and checked what writes the address, 

`D9 56 40 D9 C0 DE E2 DE F1 D9 5E 5C D9 45 10 D9 56 44`

I thought I could just nop it, it just crashes. This is where I'm stuck.

I was hoping somebody could figure this out. I can't.

**Comment by Rydian** [comment]
With floats you don't want to nop stuff, especially stuff that ends with p, because unbalancing the stack will cause issues.  This is a slightly different case, in that the game likely crashes because the draw distance is removed when exiting an area, and you're removing the code that would set it to begin with.

So I mentioned that because in both cases, if you want your own specific custom value, then you insert some sort of hardcoded overwrite (like a mov of a hex value that represents a float).  Thing is three bytes is not enough to fit this in as a replacement.  In this case we have an fld directly before the fst, which has the job of loading the value that then gets written.  Since we don't care about that happening we can theoretically replace both opcodes, so we have 6 bytes to work with (as both the fld and fst are 3 bytes each).

Problem is `mov [esi+40],47435000` (to set it to 50K in float) takes up 7 bytes...

So this can be worked around by doing some hook/injection thing (which I don't know the syntax for if it can be done here), or we can try to modify the load (the fld right above the fst) to load a large value from somewhere else in the loaded memory.

`fld dword ptr [ebp+0C]` is what it is right now.
`fld dword ptr [ebp+22]` is what we can change it to that will load an extremely large value.

So...

```
D9 45 0C D9 56 40
->
D9 45 22 D9 56 40
```

Will successfully give us hella'-long render distance with one byte replaced.

... of course, 4.219700799 x 10^20 as a render distance _is incredibly laggy_ and we can't be sure that the value at that location will stay that huge since it's a misinterpretation of other data.  Some sort of hook to insert an overwrite of the draw distance (after the fst, ignoring the fld) to 100-500K would be more preferable.

**Comment by Poshwosh** [comment]
> Some sort of hook to insert an overwrite of the draw distance (before or after the fst doesn't matter) to 100-500K would be more preferable.

I want to say yes but 100 and 500k might be a bit too low and too much respective, that's why I thought of just boosting the maximum to something modern machines can still run.

I don't really understand  why ebp+22 worked specifically though? I don't know much of assembly so.. I don't know how much an explanation would do. I appreciate your effort Rydian. 

I tend to mess as opposed to understand, and with that said you got me to mess with those values, I changed esi+40 to esi+20 , and suddenly it stopped resetting the draw distance to 15k, it worked! Pure damn luck. It doesn't seem to lag either. Do you understand what it does now by using 20 instead of 40? 

`00 C8 42 00 ?? ?? ?? 00 00 34 42 00 00 D2 44` to `00 C8 42 00 50 43 47 00 00 34 42 00 00 D2 44`
`D9 56 40 D9 C0 DE E2 DE F1 D9 5E 5C D9 45 10 D9 56 44` to `D9 56 20 D9 C0 DE E2 DE F1 D9 5E 5C D9 45 10 D9 56 44`

But problem is, I just discovered the first value doesn't even load till the character selection screen.... I think we should try a different method. Don't you have range of vision in your CE table? Does it reset?

**Comment by Rydian** [comment]
`ebp+22`, at that point, is a value that, when interpreted as a float, is really high.  It's some other value (or mix of values) used for other things elsewhere, I just checked around until I found an offset that was technically a really high value.  I don't know what those values are or where they came from, so a proper solution/overwrite is definitely suggested.

In CE, you can highlight a line of code and go to Debug -> Toggle Breakpoint.  Next time the code executes (immediately for some lines of code) the game will pause and the memory browser will display extra info.  Toggle the breakpoint again and then go to Debug -> Run to resume the game.  The top-right of the memory browser will display extra info, such as the values of the registers at that time.

So for example when `fst dword ptr [esi+40]` executes, it shows 0C16A898 as the value of `esi` at that time for me.  0C16A898+40 = 0C16A8D8, and the value at that address is 15000 (so that's where it's storing the draw distance).

A lot of my scripts work via injection to add a bit more code for overwrites and such, would need to see how this would be done in Kanan.

**Comment by Poshwosh** [comment]
> 0C16A898+40 = 0C16A8D8, and the value at that address is 15000

Gasp.

So instead of writing to 0C16A8D8 (esi+40, 15000), I made it write to 0C16A8B8, and now the 15k doesn't get effected by it, while another poor address is? I want to ask how would I just make it not write to anything but we're going a bit off topic haha.

There's still the problem of the address only being written after loading.

This was in Step29's broken mod_sharker config (which @C0ZIEST provided), maybe it could provide some help

```
;Dramatically increase render range a hundredfold (may need to move Range of Vision slider in options to take effect)
[RangeOfVisionMax]
SEARCH1= D9 ?? ?? D9 05 ?? ?? ?? ?? D8 ?? ?? DF E0 F6 C4 05 7A 05 D9 ?? ?? EB 02 DD D8 D9 EE
SEARCH2= 85 C0 7D 06 D8 05 ?? ?? ?? ?? DC 35 ?? ?? ?? ?? 53 D9 9F A4
REPLACE1=D9 ?? ?? B8 33 29 33 29 90 D8 ?? ?? DF E0 F6 C4 05 7A 05 D9 ?? ?? EB 02 DD D8 D9 EE
REPLACE2=85 C0 7D 06 D8 05 ?? ?? ?? ?? DC 0D ?? ?? ?? ?? 53 D9 9F A4
```

**Comment by cursey** [comment]
`ebp+0Ch` in `fld [ebp+0Ch]` is the second argument passed to the function. The function is called from 7 different places. If you find the one of interest (if there is only one), you may be able to just modify the argument as its passed to the function instead of modifying the function.

**Comment by Rydian** [comment]
Is there a simple way to allocate memory, jump there for additional code, and jump back?

**Comment by cursey** [comment]
`allocateMemory(len)` from `Defaults.js` will allocate a read/write/executable area of memory you can jump to and jump back from. You have to manually setup the jumps though which is kind of a pain and I could write some helper functions to do that.

**Comment by Rydian** [comment]
If you could, it would help in bringing a lot of other scripts over too (like FoV modifications) where we need to overwrite the results of floating point math.

**Comment by cursey** [comment]
Okay, so I added a helper function `insertJmp` to `Defaults.js`. You can use it alongside `allocateMemory` like so:
1. Allocate a new block of memory for our code to go into using `allocateMemory(len)` where `len` is how many bytes we need to reserve to hold all of our code including the jump out of it.
2. Use `patch(address, [0x90, ...])` to `nop` out as many instructions at the destination that is needed to have at least 5 bytes of free space (for the `jmp` to our block of memory).
3. `insertJmp(address, allocatedMemoryLocation)` to insert a jump at the location we `nop`'d to our allocated block.
4. `patch(allocatedMemoryLocation, [...])` the code in our block.
5. `insertJmp(allocatedMemoryLocation.add(codeLen), address.add(5))` to insert a jump at the end of our code block back to the original address.

I'll do a concrete example for this feature in a bit.

**Comment by cursey** [comment]
Ok so here it is. I just took what I found in this issue and turned it into a script using the method described above. It's ~not~ doing what was intended so take a look, fix, then make a pull request ^^

```
// Set this to your desired range of vision.
var desiredRangeOfVision = 50000.0;

// This is the location we will patch, it is the instruction fld [ebp+0Ch] which
// is followed by fst [esi+40h]. We want to replace both of these instructions 
// but do not have enough room to do so, so we will setup a block and jump to 
// and from it.
var thePatchLocation = scan('D9 45 0C D9 56 40');

// Our patch is going to simply move the hex representation of our floating
// point value to esi+40h.
var thePatch = [
    0xC7, 0x46, 0x40, 0xFF, 0xFF, 0xFF, 0xFF // mov [esi+40h], 0xFFFFFFFF
];

// We allocate 5 extra bytes then necessary for the jmp out of our block.
var ourCodeLocation = allocateMemory(thePatch.length + 5);

dmsg(ourCodeLocation);
patch(ourCodeLocation, thePatch);

// Replace the placeholder 0xFFFFFFFF with the desired range of vision.
var buf = new ArrayBuffer(4);
var bytes = new Uint8Array(buf);
var floats = new Float32Array(buf);

floats[0] = desiredRangeOfVision;

patch(ourCodeLocation.add(3), Array.from(bytes));

// Insert the jmp at the end of our code block back to the destinations next 
// instruction.
insertJmp(ourCodeLocation.add(thePatch.length), thePatchLocation.add(6));

// Now our code block is all setup, just make the jmp at thePatchLocation to it.
patch(thePatchLocation, Array(6).fill(0x90)); // 2 instructions take up 6 bytes.
insertJmp(thePatchLocation, ourCodeLocation);
```

You will need the latest `Defaults.js` for this. The inserting jumps and code block allocation all works, it just isn't extending the range of vision (the opposite happens really, nothing seems to get rendered).

**Comment by Rydian** [comment]
Some sort of alignment issue or something.  The injected code looks like this _initially_...

```
0A850000 - C7 46 40 0024F448     - mov [esi+40],48F42400
0A850007 - E9 46906FF6           - jmp Client.exe+B49052
```

But trying to view it in context shows it being interpreted as this...

```
0A84FFFF - 00 C7                 - add bh,al
0A850001 - 46                    - inc esi
0A850002 - 40                    - inc eax
0A850003 - 00 24 F4              - add [esp+esi*8],ah
0A850006 - 48                    - dec eax
0A850007 - E9 46906FF6           - jmp Client.exe+B49052
```

Due to directly following a bunch of other bytes that may or may not constitute an entire opcode.

I don't have enough experience to say how to avoid that, though CE seems to allocate new memory after unallocated memory (and not specifically at the end).

**Comment by Rydian** [comment]
I tried some more stuff and there's definitely other data getting written to the new location, not just thePatch.  I padded thePatch with a bunch of nops in a row to see if things would survive.

```
var thePatch = [
    0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0x90, 0xC7, 0x46, 0x40, 0x00, 0x24, 0xF4, 0x48
];
```

So that's 12 nops added before the mov.  The jmp the script inserts to go to the code is set to jump to `1FD40000`.

But `1FD40000` actually looks like this...

```
1FD40000 - 90                    - nop 
1FD40001 - 90                    - nop 
1FD40002 - 90                    - nop 
1FD40003 - 00 24 F4              - add [esp+esi*8],ah
1FD40006 - 48                    - dec eax
1FD40007 - 90                    - nop 
1FD40008 - 90                    - nop 
1FD40009 - 90                    - nop 
1FD4000A - 90                    - nop 
1FD4000B - 90                    - nop 
1FD4000C - C7 46 40 0024F448     - mov [esi+40],48F42400 { 1223959552 }
1FD40013 - E9 3A9020E1           - jmp Client.exe+B49052
```

**Comment by cursey** [comment]
That's because it's trying to patch in the `desiredRangeOfVision` at `ourCodeLocation + 3`. From the following code:

```
// Replace the placeholder 0xFFFFFFFF with the desired range of vision.
var buf = new ArrayBuffer(4);
var bytes = new Uint8Array(buf);
var floats = new Float32Array(buf);

floats[0] = desiredRangeOfVision;

patch(ourCodeLocation.add(3), Array.from(bytes));
```

As far as I know, VirtualAlloc always returns an address aligned to a page boundary so alignment shouldn't be an issue.

**Comment by Rydian** [comment]
Ah yup I when I padded I didn't account for the changing offset, so I've removed that bit for testing.

Adding the `fld` and `fst` to the injected code (in case they _were_ actually important) made an improvement in that the ground mesh of an area (with the modified render distance) and NPCs will show, but no decorations render (houses and such).

So I decided to compare.  I changed thePatch to...
`0xD9, 0x45, 0x0C, 0xD9, 0x56, 0x40, 0xC7, 0x46, 0x40, 0x00, 0x24, 0xF4, 0x48`
To make Kana's injection match CE's, at least with the written bytes up to the jump back.

Unmodified code: http://i.imgur.com/3jZtXFv.png

CE's hook: http://i.imgur.com/MnMCBC2.png
Kanan's hook: http://i.imgur.com/GEpxCwW.png

CE's injection: http://i.imgur.com/LHlzCXJ.png
Kanan's injection: http://i.imgur.com/wfamqYe.png

I thought maybe it was some register changing when it's not supposed to, but nothing other than EIP changes between the `fst` right before the jump and the `mov` overwriting the result.  And the game appears to jump there and back just fine.
http://i.imgur.com/oWLKxAN.png

So I'm still too inexperienced to troubleshoot this.

EDIT: gsfjlhsjlkhfaskjh I just realized I should have tested it alone earlier.   **It works just fine if all other game mod scripts are disabled.**  So it's some sort of conflict.

**Comment by Rydian** [comment]
So the following version of the script works...

```
// Set this to your desired range of vision.
var desiredRangeOfVision = 500000.0;

// This is the location we will patch, it is the instruction fld [ebp+0Ch] which
// is followed by fst [esi+40h]. We want to replace both of these instructions 
// but do not have enough room to do so, so we will setup a block and jump to 
// and from it.
var thePatchLocation = scan('D9 45 0C D9 56 40');

// Our patch is going to simply move the hex representation of our floating
// point value to esi+40h.
var thePatch = [
    0xD9, 0x45, 0x0C, 0xD9, 0x56, 0x40, 0xC7, 0x46, 0x40, 0x00, 0x24, 0xF4, 0x48
];

// We allocate 5 extra bytes then necessary for the jmp out of our block.
var ourCodeLocation = allocateMemory(thePatch.length + 5);

dmsg(ourCodeLocation);
patch(ourCodeLocation, thePatch);

// Replace the placeholder 0xFFFFFFFF with the desired range of vision.
var buf = new ArrayBuffer(4);
var bytes = new Uint8Array(buf);
var floats = new Float32Array(buf);

floats[0] = desiredRangeOfVision;

//patch(ourCodeLocation.add(3), Array.from(bytes));

// Insert the jmp at the end of our code block back to the destinations next 
// instruction.
insertJmp(ourCodeLocation.add(thePatch.length), thePatchLocation.add(6));

// Now our code block is all setup, just make the jmp at thePatchLocation to it.
patch(thePatchLocation, Array(6).fill(0x90)); // 2 instructions take up 6 bytes.
insertJmp(thePatchLocation, ourCodeLocation);
```

With the contents of disabled.txt as follows.

```
AlwaysTransCollectMode.js
BitmapFont.js
BorderlessWindowedMode.js
ChatAllowSameMsg.js
ChatNoRateLimit.js
ClientSideDevCat.js
ColorAltText.js
DisableScreenFlash.js
DllLoader.js
DontTargetNPCs.js
EnableInterfaceWhileDead.js
FarAltText.js
FighterNoForce.js
FreeIndoorCamera.js
FreezeTimeOfDay.js
HideMainTitle.js
HideNPCCurtains.js
HideObjects.js
HideSecondTitle.js
HotkeyAnyItem.js
IgnoreTranslucency.js
LargerTTFontResize.js
ManaTunnelLagFix.js
MiniTitleMenu.js
NoPersonalShopCreationMsg.js
NoSkillRankUpWindow.js
OneClickRevive.js
PartyChatMarketFilter.js
PatternScanSnapshot.js
RangedAttackSwap.js
RemoveScreenShake.js
ShowDeadlyHP.js
ShowItemTradePrice.js
ZeroFogDistance.js
```

Using this (latest as of the time of this post) version of Kanan.
https://github.com/cursey/kanan/tree/dfe4e85b7e4e349331b5f553627c2009c018a914

I ran Kanan with all scripts enabled and it had the issue you and I kept running into, so it's definitely some other script doing it.  I've just been up all night and I don't feel like restarting the client 10 more times to figure it out.

**Comment by Poshwosh** [comment]
I attempted to run the script with the recent kanan build with no changes to disabled.txt and it worked (really well too, you two are amazing!!), so I don't see what's the problem? 

Is it the alignment? While I don't know much in terms of assembly, your comparison pictures seem identical, just jumping into different addresses which contain the same values.

> Adding the fld and fst to the injected code (in case they were actually important) made an improvement in that the ground mesh of an area (with the modified render distance) and NPCs will show, but no decorations render (houses and such).

This sounds exactly like what HideObjects.js does, maybe that's the problem.

Lastly I think by default the value should be 30000, I actually ran into some FPS drops below 60 around north Dunbarton with 50k.
